### PR TITLE
add catalog controller and curated collection of MQTT chart

### PR DESCRIPTION
**Pull request by shanewxy**
Related issue: https://github.com/cnrancher/octopus/issues/46

### CODE REVIEW

File: pkg/controllers/catalog/controller.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	Name            = "catalog-controller"
+	namespace       = "kube-system"
+	MQTTCatalogName = "mqtt"
+	MQTTCatalogURL  = "http://charts.cnrancher.cn"
```
Comment (guangbochen) [inline pkg/controllers/catalog/controller.go]: set this to `http://charts.cnrancher.cn/mqtt`

File: pkg/controllers/catalog/controller.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/rancher/wrangler/pkg/apply"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	Name            = "catalog-controller"
+	namespace       = "kube-system"
+	MQTTCatalogName = "mqtt"
```
Comment (guangbochen) [inline pkg/controllers/catalog/controller.go]: rename it to `mqtt-library`

File: pkg/controllers/catalog/deploy/emqx.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+apiVersion: helm.cattle.io/v1
+kind: HelmChart
+metadata:
+  name: my-mqtt
+  namespace: kube-system
+spec:
+  chart: emqx
+  repo: http://charts.cnrancher.cn
```
Comment (guangbochen) [inline pkg/controllers/catalog/deploy/emqx.yaml]: change to http://charts.cnrancher.cn/mqtt

File: pkg/steve/pkg/catalogapi/catalogapi.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func (s *Server) Setup(ctx context.Context, server *steveserver.Server) error {
+	s.ctx = ctx
+	s.asl = server.AccessSetLookup
+	s.auth = server.AuthMiddleware
+	s.cf = server.ClientFactory
+	s.schemas = server.BaseSchemas
+	store := proxy.NewProxyStore(s.cf, s.asl)
+	refreshHandler := customization.NewFooHandler()
```
Comment (guangbochen) [inline pkg/steve/pkg/catalogapi/catalogapi.go]: remove the `refreshHandler` for now as they are not working.

File: pkg/steve/pkg/catalogapi/store.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		for _, catalog := range catalogList.Items {
+			catalogs = append(catalogs, catalog)
+		}
+	}
+	for _, catalog := range catalogs {
+		if err := s.refreshCatalog(&catalog); err != nil {
+			return object, err
+		}
+	}
+	return s.Store.Create(apiOp, schema, data)
```
Comment (guangbochen) [inline pkg/steve/pkg/catalogapi/store.go]: return `data, nil` instead

**Review by guangbochen — CHANGES_REQUESTED** [review]
Can u please help to either remove foo or replace foo with catalog, thanks

**Comment by shanewxy** [comment]
updated @guangbochen

**Comment by guangbochen** [comment]
LGTM, thanks

**Pull request merged**